File: cmd/ambience/web/effects/campfire.js
'use strict';
(function (api) {
	const { makeRNG, jitterInt, clamp01, hslToRGB, positiveMod } = api._helpers;

	const DEFAULTS = {
		intro_dur: 45,
		intro_glow: 0.14,
		ending_dur: 60,
		ending_linger: 24,
		ending_glow: 0.08,
		flame_height: 14,
		flame_width: 10,
		flame_speed: 0.12,
		flicker: 0.72,
		ember_rate: 0.26,
		ember_speed: 0.62,
		glow: 0.54,
		hue: 24,
		hue_sp: 18,
		sat: 0.82,
		lmin: 0.32,
		lmax: 0.94,
		crackle_p: 0,
		lull_p: 0,
		crackle_dur: 36,
		crackle_mult: 1.85,
		lull_dur: 68,
		lull_mult: 0.55,
	};

	function applyDefaults(cfg) {
		const c = Object.assign({}, DEFAULTS, cfg || {});
		if (c.intro_dur <= 0) c.intro_dur = DEFAULTS.intro_dur;
		c.intro_glow = clamp01(c.intro_glow);
		if (c.ending_dur <= 0) c.ending_dur = DEFAULTS.ending_dur;
		if (c.ending_linger < 0) c.ending_linger = 0;
		c.ending_glow = clamp01(c.ending_glow);
		if (c.flame_height <= 0) c.flame_height = DEFAULTS.flame_height;
		if (c.flame_width <= 0) c.flame_width = DEFAULTS.flame_width;
		if (c.flame_speed <= 0) c.flame_speed = DEFAULTS.flame_speed;
		if (c.flicker <= 0) c.flicker = DEFAULTS.flicker;
		if (c.ember_rate <= 0) c.ember_rate = DEFAULTS.ember_rate;
		if (c.ember_speed <= 0) c.ember_speed = DEFAULTS.ember_speed;
		if (c.glow <= 0) c.glow = DEFAULTS.glow;
		if (c.hue === 0) c.hue = DEFAULTS.hue;
		if (c.hue_sp < 0) c.hue_sp = 0;
		if (c.sat <= 0) c.sat = DEFAULTS.sat;
		if (c.lmin <= 0) c.lmin = DEFAULTS.lmin;
		if (c.lmax <= 0) c.lmax = DEFAULTS.lmax;
		if (c.lmax < c.lmin) [c.lmin, c.lmax] = [c.lmax, c.lmin];
		if (c.crackle_dur <= 0) c.crackle_dur = DEFAULTS.crackle_dur;
		if (c.crackle_mult <= 0) c.crackle_mult = DEFAULTS.crackle_mult;
		if (c.lull_dur <= 0) c.lull_dur = DEFAULTS.lull_dur;
		if (c.lull_mult <= 0) c.lull_mult = DEFAULTS.lull_mult;
		return c;
	}

	class Campfire {
		constructor(w, h, cfg, seed) {
			this.w = w;
			this.h = h;
			this.grid = new Uint8ClampedArray(w * h * 3);
			this.seed = Number(seed || Date.now());
			this.tick = 0;
			this.timers = {};
			this.values = {};
			this.cfg = applyDefaults(cfg);
		}

		setConfig(cfg) {
			this.cfg = applyDefaults(Object.assign({}, this.cfg, cfg));
		}

		restoreSnapshot(snap) {
			const state = snap.state || snap;
			this.setConfig(snap.config || {});
			this.tick = state.tick || snap.tick || 0;
			this.timers = Object.assign({}, state.timers || {});
			this.values = Object.assign({}, state.values || {});
			if (typeof snap.seed === 'number') this.seed = snap.seed;
			if (snap.gridW > 0 && snap.gridH > 0) {
				this.w = snap.gridW;
				this.h = snap.gridH;
			}
		}

		_eventRng(salt) {
			return makeRNG(((this.seed >>> 0) ^ (((this.tick + salt) * 2654435761) >>> 0)) >>> 0);
		}

		_hash(index) {
			const x = Math.sin((this.seed * 0.000001 + index * 12.9898) * 43758.5453);
			return x - Math.floor(x);
		}

		_phaseProgress(total, left) {
			if (left <= 1 || total <= 1) return 1;
			const elapsed = total - left;
			if (elapsed <= 0) return 0;
			return clamp01(elapsed / Math.max(1, total - 1));
		}

		_fillCell(ctx, sx, sy, ceilSx, ceilSy, x, y, w, h, color, alpha) {
			ctx.fillStyle = color;
			ctx.globalAlpha = alpha == null ? 1 : alpha;
			ctx.fillRect(Math.floor(x * sx), Math.floor(y * sy), Math.max(1, Math.ceil(w * sx || ceilSx)), Math.max(1, Math.ceil(h * sy || ceilSy)));
			ctx.globalAlpha = 1;
		}

		triggerEvent(name) {
			const rng = this._eventRng(name.length + 67);
			switch (name) {
				case 'crackle':
					this.timers.crackle = jitterInt(rng, this.cfg.crackle_dur, 0.3);
					this.values.crackle_gain = this.cfg.crackle_mult * (0.75 + rng() * 0.5);
					return true;
				case 'lull':
					this.timers.lull = jitterInt(rng, this.cfg.lull_dur, 0.3);
					return true;
				case 'intro':
					this.timers.crackle = 0;
					this.timers.lull = 0;
					this.timers.ending = 0;
					this.values.crackle_gain = 1;
					this.timers.intro = Math.max(1, Math.round(this.cfg.intro_dur));
					this.values.intro_total = this.timers.intro;
					return true;
				case 'ending':
					this.timers.intro = 0;
					this.timers.crackle = 0;
					this.timers.lull = 0;
					this.values.crackle_gain = 1;
					this.timers.ending = Math.max(1, Math.round(this.cfg.ending_dur + Math.max(0, this.cfg.ending_linger)));
					this.values.ending_total = this.timers.ending;
					return true;
			}
			return false;
		}

		step() {
			this.tick++;
			for (const key of Object.keys(this.timers)) {
				if (this.timers[key] > 0) this.timers[key]--;
			}
			if (!this.timers.crackle || this.timers.crackle <= 0) this.values.crackle_gain = 1;
			api._helpers.paintProceduralGrid(this);
		}

		_flameLevel() {
			let level = 1;
			if (this.timers.crackle > 0) level *= this.values.crackle_gain || this.cfg.crackle_mult;
			if (this.timers.lull > 0) level *= this.cfg.lull_mult;
			if (this.timers.intro > 0) {
				const total = this.values.intro_total || this.cfg.intro_dur;
				const progress = this._phaseProgress(total, this.timers.intro);
				level *= this.cfg.intro_glow + (1 - this.cfg.intro_glow) * progress;
			}
			if (this.timers.ending > 0) {
				const total = this.values.ending_total || (this.cfg.ending_dur + this.cfg.ending_linger);
				const progress = this._phaseProgress(total, this.timers.ending);
				level *= 1 - (1 - this.cfg.ending_glow) * progress;
			}
			return Math.max(0.05, level);
		}

		render(ctx, canvasW, canvasH, opts) {
			api._helpers.renderPixelGridEffect(this, ctx, canvasW, canvasH, opts);
		}
	}

	api.presets['campfire'] = [
		{
			key: 'small-fire',
			label: 'small fire',
			config: {
				flame_height: 9,
				flame_width: 7,
				flame_speed: 0.1,
				flicker: 0.56,
				ember_rate: 0.18,
				ember_speed: 0.52,
				glow: 0.4,
				hue: 22,
				hue_sp: 12,
				sat: 0.76,
				lmin: 0.28,
				lmax: 0.88,
			},
		},
		{
			key: 'steady-campfire',
			label: 'steady campfire',
			config: {
				flame_height: 14,
				flame_width: 10,
				flame_speed: 0.12,
				flicker: 0.72,
				ember_rate: 0.26,
				ember_speed: 0.62,
				glow: 0.54,
				hue: 24,
				hue_sp: 18,
				sat: 0.82,
				lmin: 0.32,
				lmax: 0.94,
				crackle_p: 0.0008,
			},
		},
		{
			key: 'crackling-fire',
			label: 'crackling fire',
			config: {
				flame_height: 16,
				flame_width: 11,
				flame_speed: 0.15,
				flicker: 0.92,
				ember_rate: 0.34,
				ember_speed: 0.78,
				glow: 0.62,
				hue: 21,
				hue_sp: 22,
				sat: 0.88,
				lmin: 0.34,
				lmax: 0.96,
				crackle_p: 0.0015,
				crackle_mult: 2.15,
				crackle_dur: 48,
			},
		},
		{
			key: 'late-embers',
			label: 'late embers',
			config: {
				intro_glow: 0.1,
				ending_glow: 0.14,
				flame_height: 8,
				flame_width: 8,
				flame_speed: 0.08,
				flicker: 0.42,
				ember_rate: 0.3,
				ember_speed: 0.48,
				glow: 0.34,
				hue: 18,
				hue_sp: 14,
				sat: 0.68,
				lmin: 0.24,
				lmax: 0.8,
				lull_p: 0.0014,
				lull_mult: 0.42,
			},
		},
	];
	api.effects['campfire'] = Campfire;
})(window.AmbienceSim);

File: sim/campfire.go
package sim

import (
	"fmt"
	"math"
	"sync"

	"github.com/nelsong6/ambience/rngutil"
)

type CampfireConfig = ProceduralConfig
type CampfireState = ProceduralState
type CampfireSnapshot = ProceduralSnapshot
type CampfirePersistedState = ProceduralPersistedState

// Campfire is a dedicated sim type for the campfire effect.
type Campfire struct {
	mu sync.Mutex

	W, H int
	Grid [][]Pixel

	rng    *rngutil.RNG
	cfg    CampfireConfig
	tick   int
	timers map[string]int
	values map[string]float64
	log    []LogEntry
}

var campfireDefaultsLocal = CampfireConfig{
	"intro_dur":     45,
	"intro_glow":    0.14,
	"ending_dur":    60,
	"ending_linger": 24,
	"ending_glow":   0.08,
	"flame_height":  14.0,
	"flame_width":   10.0,
	"flame_speed":   0.12,
	"flicker":       0.72,
	"ember_rate":    0.26,
	"ember_speed":   0.62,
	"glow":          0.54,
	"hue":           24,
	"hue_sp":        18,
	"sat":           0.82,
	"lmin":          0.32,
	"lmax":          0.94,
	"crackle_p":     0.0,
	"lull_p":        0.0,
	"crackle_dur":   36,
	"crackle_mult":  1.85,
	"lull_dur":      68,
	"lull_mult":     0.55,
}

func CampfireSchema() EffectSchema {
	return EffectSchema{
		Name: "campfire",
		Knobs: []Knob{
			{Key: "intro_dur", Label: "intro dur", Slot: SlotSpawn, Group: "introduction", Type: KnobInt, Min: 10, Max: 180, Step: 5, Default: 45, Trigger: "intro",
				Description: "Ticks spent catching from a small glow into a stable flame."},
			{Key: "intro_glow", Label: "intro glow", Slot: SlotSpawn, Group: "introduction", Type: KnobFloat, Min: 0.05, Max: 0.5, Step: 0.01, Default: 0.14,
				Description: "Starting fraction of the final campfire intensity before the flame catches fully."},
			{Key: "ending_dur", Label: "ending dur", Slot: SlotEnd, Group: "ending", Type: KnobInt, Min: 10, Max: 220, Step: 5, Default: 60, Trigger: "ending",
				Description: "Ticks spent collapsing the flame down toward ember glow."},
			{Key: "ending_linger", Label: "ending linger", Slot: SlotEnd, Group: "ending", Type: KnobInt, Min: 0, Max: 160, Step: 5, Default: 24,
				Description: "Extra ember time after the flame has mostly died back."},
			{Key: "ending_glow", Label: "ending glow", Slot: SlotEnd, Group: "ending", Type: KnobFloat, Min: 0.02, Max: 0.35, Step: 0.01, Default: 0.08,
				Description: "Residual flame and ember intensity near the end of the outro."},
			{Key: "flame_height", Label: "flame height", Slot: SlotLever, Group: "flame", Type: KnobFloat, Min: 6, Max: 24, Step: 0.5, Default: 14,
				Description: "Overall flame height above the logs."},
			{Key: "flame_width", Label: "flame width", Slot: SlotLever, Group: "flame", Type: KnobFloat, Min: 4, Max: 18, Step: 0.5, Default: 10,
				Description: "Width of the flame body and ember bed."},
			{Key: "flame_speed", Label: "flame speed", Slot: SlotLever, Group: "flame", Type: KnobFloat, Min: 0.04, Max: 0.3, Step: 0.01, Default: 0.12,
				Description: "How quickly the flame shape flickers and rolls."},
			{Key: "flicker", Label: "flicker", Slot: SlotLever, Group: "flame", Type: KnobFloat, Min: 0.1, Max: 1.4, Step: 0.02, Default: 0.72,
				Description: "Side-to-side and height variation inside the flame body."},
			{Key: "ember_rate", Label: "ember rate", Slot: SlotLever, Group: "embers", Type: KnobFloat, Min: 0.05, Max: 0.7, Step: 0.01, Default: 0.26,
				Description: "How many embers rise from the fire during the steady state."},
			{Key: "ember_speed", Label: "ember speed", Slot: SlotLever, Group: "embers", Type: KnobFloat, Min: 0.1, Max: 1.4, Step: 0.02, Default: 0.62,
				Description: "How quickly embers travel upward and fade."},
			{Key: "glow", Label: "glow", Slot: SlotLever, Group: "embers", Type: KnobFloat, Min: 0.05, Max: 0.9, Step: 0.01, Default: 0.54,
				Description: "Strength of the warm localized light cast around the campfire."},
			{Key: "hue", Label: "hue", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 8, Max: 50, Step: 1, Default: 24,
				Description: "Base flame hue. Lower values lean redder; higher values lean more yellow-orange."},
			{Key: "hue_sp", Label: "hue spread", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0, Max: 28, Step: 1, Default: 18,
				Description: "Variation between coal reds, orange mids, and bright flame tips."},
			{Key: "sat", Label: "saturation", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0.2, Max: 1, Step: 0.01, Default: 0.82,
				Description: "Overall color saturation of the fire and ember tones."},
			{Key: "lmin", Label: "light min", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0.1, Max: 0.7, Step: 0.01, Default: 0.32,
				Description: "Minimum lightness used for darker coals, logs, and outer flame edges."},
			{Key: "lmax", Label: "light max", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0.4, Max: 1, Step: 0.01, Default: 0.94,
				Description: "Maximum lightness used for the hottest flame cores and bright embers."},
			{Key: "crackle_p", Label: "crackle", Slot: SlotEvent, Type: KnobFloat, Min: 0, Max: 0.02, Step: 0.0005, Default: 0, Trigger: "crackle",
				Description: "Per-tick chance of a brighter crackle that throws extra embers."},
			{Key: "lull_p", Label: "lull", Slot: SlotEvent, Type: KnobFloat, Min: 0, Max: 0.02, Step: 0.0005, Default: 0, Trigger: "lull",
				Description: "Per-tick chance of the flame briefly settling into a lower, calmer burn."},
			{Key: "crackle_dur", Label: "crackle dur", Slot: SlotEventMod, Group: "crackle", Type: KnobInt, Min: 8, Max: 160, Step: 4, Default: 36,
				Description: "Duration of the brighter crackling burst."},
			{Key: "crackle_mult", Label: "crackle x", Slot: SlotEventMod, Group: "crackle", Type: KnobFloat, Min: 1.05, Max: 3, Step: 0.05, Default: 1.85,
				Description: "Intensity multiplier applied while a crackle burst is active."},
			{Key: "lull_dur", Label: "lull dur", Slot: SlotEventMod, Group: "lull", Type: KnobInt, Min: 10, Max: 220, Step: 5, Default: 68,
				Description: "Duration of the quieter lower-flame window."},
			{Key: "lull_mult", Label: "lull x", Slot: SlotEventMod, Group: "lull", Type: KnobFloat, Min: 0.1, Max: 1, Step: 0.05, Default: 0.55,
				Description: "Flame intensity multiplier applied while lull is active."},
		},
	}
}

func defaultCampfireConfig() CampfireConfig { return cloneConfig(campfireDefaultsLocal) }

func mergeCampfireDefaults(cfg CampfireConfig) CampfireConfig {
	out := defaultCampfireConfig()
	for k, v := range cfg {
		out[k] = v
	}
	if out["intro_dur"] <= 0 {
		out["intro_dur"] = campfireDefaultsLocal["intro_dur"]
	}
	out["intro_glow"] = clamp01(out["intro_glow"])
	if out["ending_dur"] <= 0 {
		out["ending_dur"] = campfireDefaultsLocal["ending_dur"]
	}
	if out["ending_linger"] < 0 {
		out["ending_linger"] = 0
	}
	out["ending_glow"] = clamp01(out["ending_glow"])
	if out["flame_height"] <= 0 {
		out["flame_height"] = campfireDefaultsLocal["flame_height"]
	}
	if out["flame_width"] <= 0 {
		out["flame_width"] = campfireDefaultsLocal["flame_width"]
	}
	if out["flame_speed"] <= 0 {
		out["flame_speed"] = campfireDefaultsLocal["flame_speed"]
	}
	if out["flicker"] <= 0 {
		out["flicker"] = campfireDefaultsLocal["flicker"]
	}
	if out["ember_rate"] <= 0 {
		out["ember_rate"] = campfireDefaultsLocal["ember_rate"]
	}
	if out["ember_speed"] <= 0 {
		out["ember_speed"] = campfireDefaultsLocal["ember_speed"]
	}
	if out["glow"] <= 0 {
		out["glow"] = campfireDefaultsLocal["glow"]
	}
	if out["hue"] == 0 {
		out["hue"] = campfireDefaultsLocal["hue"]
	}
	if out["hue_sp"] < 0 {
		out["hue_sp"] = 0
	}
	if out["sat"] <= 0 {
		out["sat"] = campfireDefaultsLocal["sat"]
	}
	if out["lmin"] <= 0 {
		out["lmin"] = campfireDefaultsLocal["lmin"]
	}
	if out["lmax"] <= 0 {
		out["lmax"] = campfireDefaultsLocal["lmax"]
	}
	if out["lmax"] < out["lmin"] {
		out["lmin"], out["lmax"] = out["lmax"], out["lmin"]
	}
	if out["crackle_dur"] <= 0 {
		out["crackle_dur"] = campfireDefaultsLocal["crackle_dur"]
	}
	if out["crackle_mult"] <= 0 {
		out["crackle_mult"] = campfireDefaultsLocal["crackle_mult"]
	}
	if out["lull_dur"] <= 0 {
		out["lull_dur"] = campfireDefaultsLocal["lull_dur"]
	}
	if out["lull_mult"] <= 0 {
		out["lull_mult"] = campfireDefaultsLocal["lull_mult"]
	}
	return out
}

func NewCampfire(w, h int, seed int64, cfg CampfireConfig) *Campfire {
	grid := make([][]Pixel, h)
	for i := range grid {
		grid[i] = make([]Pixel, w)
	}
	return &Campfire{
		W:      w,
		H:      h,
		Grid:   grid,
		rng:    rngutil.New(seed),
		cfg:    mergeCampfireDefaults(cfg),
		timers: make(map[string]int),
		values: make(map[string]float64),
	}
}

func (c *Campfire) Resize(w, h int) {
	if w <= 0 || h <= 0 {
		return
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	if w == c.W && h == c.H {
		return
	}
	c.W = w
	c.H = h
	c.Grid = make([][]Pixel, h)
	for i := range c.Grid {
		c.Grid[i] = make([]Pixel, w)
	}
}

func (c *Campfire) SetConfig(cfg CampfireConfig) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.cfg = mergeCampfireDefaults(cfg)
}

func (c *Campfire) EffectiveConfig() CampfireConfig {
	c.mu.Lock()
	defer c.mu.Unlock()
	return cloneConfig(c.cfg)
}

func (c *Campfire) Snapshot() CampfireSnapshot {
	c.mu.Lock()
	defer c.mu.Unlock()
	return CampfireSnapshot{
		ProceduralState: c.snapshotStateLocked(),
		RNGState:        c.rng.State(),
	}
}

func (c *Campfire) RestoreSnapshot(snap CampfireSnapshot) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.restoreStateLocked(snap.ProceduralState)
	if snap.RNGState != 0 {
		c.rng.SetState(snap.RNGState)
	}
}

func (c *Campfire) SnapshotPersistedState() CampfirePersistedState {
	c.mu.Lock()
	defer c.mu.Unlock()
	return CampfirePersistedState{
		ProceduralState: c.snapshotStateLocked(),
		RNGState:        c.rng.State(),
	}
}

func (c *Campfire) RestorePersistedState(ps CampfirePersistedState) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.restoreStateLocked(ps.ProceduralState)
	if ps.RNGState != 0 {
		c.rng.SetState(ps.RNGState)
	}
}

func (c *Campfire) snapshotStateLocked() ProceduralState {
	return ProceduralState{
		Tick:   c.tick,
		Timers: cloneTimerMap(c.timers),
		Values: cloneValueMap(c.values),
	}
}

func (c *Campfire) restoreStateLocked(state ProceduralState) {
	c.tick = state.Tick
	c.timers = cloneTimerMap(state.Timers)
	if c.timers == nil {
		c.timers = make(map[string]int)
	}
	c.values = cloneValueMap(state.Values)
	if c.values == nil {
		c.values = make(map[string]float64)
	}
}

func (c *Campfire) CurrentTick() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.tick
}

func (c *Campfire) PerturbRNG(delta int64) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.rng.Mix(delta)
}

func (c *Campfire) DrainLog() []LogEntry {
	c.mu.Lock()
	defer c.mu.Unlock()
	if len(c.log) == 0 {
		return nil
	}
	out := c.log
	c.log = nil
	return out
}

func (c *Campfire) appendLog(kind, desc string) {
	c.log = append(c.log, LogEntry{Tick: c.tick, Type: kind, Desc: desc})
	if len(c.log) > 200 {
		c.log = c.log[len(c.log)-200:]
	}
}

func (c *Campfire) intCfg(key string) int {
	return int(math.Round(c.cfg[key]))
}

func (c *Campfire) TriggerEvent(name string) bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	switch name {
	case "crackle":
		c.startCrackleLocked("triggered")
	case "lull":
		c.startLullLocked("triggered")
	case "intro":
		c.startIntroLocked()
		c.appendLog("intro", fmt.Sprintf("started (dur=%d, glow=%.2f)", c.timers["intro"], c.cfg["intro_glow"]))
	case "ending":
		c.startEndingLocked()
		c.appendLog("ending", fmt.Sprintf("started (fade=%d, linger=%d)", c.intCfg("ending_dur"), c.intCfg("ending_linger")))
	default:
		return false
	}
	return true
}

func (c *Campfire) Step() {
	c.mu.Lock()
	defer c.mu.Unlock()

	c.tick++
	for key, value := range c.timers {
		if value > 0 {
			c.timers[key] = value - 1
		}
	}
	c.stepLocked()
}

func (c *Campfire) startCrackleLocked(verb string) {
	c.timers["crackle"] = jitterInt(c.rng, c.intCfg("crackle_dur"), 0.3)
	c.values["crackle_gain"] = c.cfg["crackle_mult"] * (0.75 + c.rng.Float64()*0.50)
	c.appendLog("crackle", fmt.Sprintf("%s (dur=%d, x%.2f)", verb, c.timers["crackle"], c.values["crackle_gain"]))
}

func (c *Campfire) startLullLocked(verb string) {
	c.timers["lull"] = jitterInt(c.rng, c.intCfg("lull_dur"), 0.3)
	c.appendLog("lull", fmt.Sprintf("%s (dur=%d, x%.2f)", verb, c.timers["lull"], c.cfg["lull_mult"]))
}

func (c *Campfire) startIntroLocked() {
	c.timers["crackle"] = 0
	c.timers["lull"] = 0
	c.timers["ending"] = 0
	c.values["crackle_gain"] = 1
	c.timers["intro"] = c.intCfg("intro_dur")
	c.values["intro_total"] = float64(c.timers["intro"])
}

func (c *Campfire) startEndingLocked() {
	c.timers["intro"] = 0
	c.timers["crackle"] = 0
	c.timers["lull"] = 0
	c.values["crackle_gain"] = 1
	endingTotal := c.intCfg("ending_dur") + max(0, c.intCfg("ending_linger"))
	if endingTotal < 1 {
		endingTotal = max(1, c.intCfg("ending_dur"))
	}
	c.timers["ending"] = endingTotal
	c.values["ending_total"] = float64(endingTotal)
}

func (c *Campfire) stepLocked() {
	if c.timers["crackle"] <= 0 {
		c.values["crackle_gain"] = 1
	}
	if c.timers["intro"] <= 0 {
		delete(c.values, "intro_total")
	}
	if c.timers["ending"] <= 0 {
		delete(c.values, "ending_total")
	}
	if c.timers["intro"] > 0 || c.timers["ending"] > 0 {
		return
	}
	if c.timers["crackle"] <= 0 && c.cfg["crackle_p"] > 0 && c.rng.Float64() < c.cfg["crackle_p"] {
		c.startCrackleLocked("started")
	}
	if c.timers["lull"] <= 0 && c.cfg["lull_p"] > 0 && c.rng.Float64() < c.cfg["lull_p"] {
		c.startLullLocked("started")
	}
}
